examples/machine_vision_object_pickup/evaUtilities.py

```diff
@@ -0,0 +1,145 @@
+# Automata Technologies - Andrea Antonello, 2019
+import math
+import operator
+import numpy as np
+
+
+def _convert_to_float(fl):
+    """ This method converts ONLY the numeric values of a string into floats """
+    try:
+        return float(fl)
+    except (ValueError, TypeError):
+        return fl
+
+
+def _wrap_to_pi(angle):
+    """ This method wrap the input angle to 360 [deg]
+    angle : [deg] """
+    ang2pi = angle - (angle // (2 * np.pi)) * 2 * np.pi
+    if ang2pi > np.pi or ang2pi < - np.pi:
+        ang = ang2pi - np.sign(ang2pi) * 2 * np.pi
+        return ang
+
+
+def _quaternion_multiply(quat1, quat0):
+    """ This method performs a standard quaternion multiplication
+    quat0 : [qR0, qV0], with qR0 being the real part of the quaternion
+    quat1 : [qR1, qV1], with qR1 being the real part of the quaternion """
+    w0, x0, y0, z0 = quat0
+    w1, x1, y1, z1 = quat1
+    return np.array([-x1 * x0 - y1 * y0 - z1 * z0 + w1 * w0,
+                     x1 * w0 + y1 * z0 - z1 * y0 + w1 * x0,
+                     -x1 * z0 + y1 * w0 + z1 * x0 + w1 * y0,
+                     x1 * y0 - y1 * x0 + z1 * w0 + w1 * z0], dtype=np.float64)
+
+
+def _solve_fk(eva, joints):
+    """ This method solves the forward kinematics problem and extract the results directly as an array
+    joints : joint angles in [rad]
+    pos : cartesian position, with respect to robot's origin [m]
+    orient : orientation quaternion of the end effector """
+    fk_results = eva.calc_forward_kinematics(joints)
+    pos_json = fk_results['position']
+    orient_json = fk_results['orientation']
+    pos = [pos_json['x'], pos_json['y'], pos_json['z']]
+    orient = [orient_json['w'], orient_json['x'], orient_json['y'], orient_json['z']]
+    return pos, orient
+
+
+def solve_ik_head_down(eva, guess, theta, xyz_absolute):
+    """ This method solves the inverse kinematics problem for the special case of the end-effector
+    pointing downwards, perpendicular to the ground.
+    guess : is the IK guess, a 1x6 array of joint angles in [rad]
+    theta : angular rotation of axis 6 [deg]
+    xyz_absolute : cartesian position, with respect to robot's origin [m] """
+    pos = [xyz_absolute[0], xyz_absolute[1], xyz_absolute[2]]  # [m]
+    pos_json = {'x': (pos[0]), 'y': (pos[1]), 'z': (pos[2])}  # [m]
+    orient_rel = [math.cos(np.deg2rad(theta) / 2), 0, 0, math.sin(np.deg2rad(theta) / 2)]
+    orient_abs = _quaternion_multiply([0, 0, 1, 0], orient_rel)
+    orient_json = {'w': (orient_abs[0]), 'x': (orient_abs[1]), 'y': (orient_abs[2]), 'z': (orient_abs[3])}
+    # Compute IK
+    result_ik = eva.calc_inverse_kinematics(guess, pos_json, orient_json)
+    success_ik = result_ik['ik']['result']
+    joints_ik = result_ik['ik']['joints']
+    return success_ik, joints_ik
+
+
+def _read_tcp_ip(sock, objects):
```

Comment:
This function could have a more explicit name. Overall, the TCP camera code should be abstracted in a class for ease of use.